### PR TITLE
Add pyright languageserver

### PR DESCRIPTION
**Pull request by mjlbach**
Closes https://github.com/neovim/nvim-lsp/issues/299
Depends on https://github.com/neovim/neovim/pull/12638

### CODE REVIEW

File: lua/nvim_lsp/pyright.lua
Comment (lithammer) [inline lua/nvim_lsp/pyright.lua:16]: Maybe include some more common root patterns, e.g. `setup.py`, `setup.cfg`, `pyproject.toml` and `requirements.txt`?

**Comment by mjlbach** [comment]
There are currently some issues discovering python module completion, still working through it.

File: lua/nvim_lsp/pyright.lua
Comment (mjlbach) [reply lua/nvim_lsp/pyright.lua]: Done, thanks for the suggestion!

**Comment by bryall** [comment]
@mjlbach what are the current issues you are seeing, and is there something I can help fix or look into?

**Comment by mjlbach** [comment]
Hi @bryall, there's not any issues, but the pyright won't work without implementing some additional features in neovim core. See: https://github.com/neovim/neovim/pull/12638

**Comment by mjlbach** [comment]
It's also important to note that the Microsoft Python Language server in C# will soon be no longer maintained. 

https://github.com/microsoft/python-language-server/issues/2096#issuecomment-666507951

**Comment by bryall** [comment]
i know, I am planning on switching to this LSP soon, and want to see if I can provide support to help get this in good shape. I will take a look at the issue

**Comment by mjlbach** [comment]
If you try out the above PR and the one from neovim core, let me know how it works :)

**Comment by mjlbach** [comment]
@justinmk I temporarily added a workaround (cc: @bryall) until the #299 is merged, so I believe this can be merged now.

**Comment by Seirdy** [comment]
> @justinmk I temporarily added a workaround (cc: @bryall) until the #299
> is merged, so I believe this can be merged now.

Dym until 
[neovim/neovim#12638](https://github.com/neovim/neovim/pull/12638) is 
merged?

Rohan Kumar

**Comment by mjlbach** [comment]
> @justinmk I temporarily added a workaround (cc: @bryall) until the #299 is merged, so I believe this can be merged now.
> Dym until [[neovim/neovim#12638](https://github.com/neovim/neovim/pull/12638)]([neovim/neovim#12638](https://github.com/neovim/neovim/pull/12638)) is merged? Rohan Kumar

Yes, fixed

**Comment by timbedard** [comment]
Well hot dang. Can we merge?

**Comment by mjlbach** [comment]
FYI for those wanting to use this and waiting for this to be merged, you can add what this PR does to your init.vim.

```lua
:lua << EOF
  local nvim_lsp = require('nvim_lsp')
  -- vim.lsp.set_log_level("debug")

  local on_attach = function(_, bufnr)
    vim.api.nvim_buf_set_option(bufnr, 'omnifunc', 'v:lua.vim.lsp.omnifunc')
    require'diagnostic'.on_attach()
    require'completion'.on_attach()

    -- Mappings.
    local opts = { noremap=true, silent=true }
    vim.api.nvim_buf_set_keymap(bufnr, 'n', 'gD', '<Cmd>lua vim.lsp.buf.declaration()<CR>', opts)
    vim.api.nvim_buf_set_keymap(bufnr, 'n', 'gd', '<Cmd>lua vim.lsp.buf.definition()<CR>', opts)
    vim.api.nvim_buf_set_keymap(bufnr, 'n', 'K', '<Cmd>lua vim.lsp.buf.hover()<CR>', opts)
    vim.api.nvim_buf_set_keymap(bufnr, 'n', 'gi', '<cmd>lua vim.lsp.buf.implementation()<CR>', opts)
    vim.api.nvim_buf_set_keymap(bufnr, 'n', '<C-k>', '<cmd>lua vim.lsp.buf.signature_help()<CR>', opts)
    vim.api.nvim_buf_set_keymap(bufnr, 'n', '<leader>wa', '<cmd>lua vim.lsp.buf.add_workspace_folder()<CR>', opts)
    vim.api.nvim_buf_set_keymap(bufnr, 'n', '<leader>wr', '<cmd>lua vim.lsp.buf.remove_workspace_folder()<CR>', opts)
    vim.api.nvim_buf_set_keymap(bufnr, 'n', '<leader>wl', '<cmd>lua vim.lsp.buf.list_workspace_folders()<CR>', opts)
    vim.api.nvim_buf_set_keymap(bufnr, 'n', '<leader>D', '<cmd>lua vim.lsp.buf.type_definition()<CR>', opts)
    vim.api.nvim_buf_set_keymap(bufnr, 'n', '<leader>rn', '<cmd>lua vim.lsp.buf.rename()<CR>', opts)
    vim.api.nvim_buf_set_keymap(bufnr, 'n', 'gr', '<cmd>lua vim.lsp.buf.references()<CR>', opts)
    vim.api.nvim_buf_set_keymap(bufnr, 'n', '<leader>e', '<cmd>lua vim.lsp.util.show_line_diagnostics()<CR>', opts)
    vim.api.nvim_buf_set_keymap(bufnr, 'n', ']d', '<cmd>lua vim.api.nvim_command("NextDiagnosticCycle")<CR>', opts)
    vim.api.nvim_buf_set_keymap(bufnr, 'n', '[d', '<cmd>lua vim.api.nvim_command("PrevDiagnosticCycle")<CR>', opts)
  end

  local configs = require('nvim_lsp/configs')
  --if not configs.pyright then
  configs.pyright = {
    default_config = {
      cmd = {"pyright-langserver", "--stdio"};
      filetypes = {"python"};
      root_dir = nvim_lsp.util.root_pattern(".git", "setup.py",  "setup.cfg", "pyproject.toml", "requirements.txt");
      settings = {
        analysis = { autoSearchPaths= true; };
        pyright = { useLibraryCodeForTypes = true; };
      };
      -- The following before_init function can be removed once https://github.com/neovim/neovim/pull/12638 is merged
      before_init = function(initialize_params)
        initialize_params['workspaceFolders'] = {{
          name = 'workspace',
          uri = initialize_params['rootUri']
        }}
      end
      };
  }
  --end

  local servers = {'gopls', 'rust_analyzer', 'sumneko_lua', 'tsserver', 'vimls', 'jsonls', 'html', 'ghcide', 'rnix', 'ocamllsp', 'pyright'}
  for _, lsp in ipairs(servers) do
    nvim_lsp[lsp].setup {
      on_attach = on_attach,
    }
  end
EOF
```

**Comment by mjlbach** [comment]
@h-michael Can we merge this?

**Comment by h-michael** [comment]
@mjlbach Could you chane module name?

**Comment by mjlbach** [comment]
From pyright?

**Comment by mjlbach** [comment]
Oh, I see. Yep, I'll fix

**Comment by h-michael** [comment]
@mjlbach See this PR https://github.com/neovim/nvim-lspconfig/pull/348

**Comment by mjlbach** [comment]
Done, I can squash commits if necessary

**Review by h-michael — APPROVED** [review]

**Comment by h-michael** [comment]
@mjlbach Thanks

**Comment by kamalmarhubi** [comment]
@mjlbach excited to try this. does it still need https://github.com/neovim/neovim/pull/12638 in order to work? if so, 🤞 that gets merged soon!

**Comment by mjlbach** [comment]
@kamalmarhubi Nope, I added a temporary fix to bypass the need for that PR, I'll remove the fix once that PR is merged.

**Comment by ellisonleao** [comment]
really dumb question @mjlbach :

is pyright supposed to work as an alternative to pyls now?

**Comment by RonanMacF** [comment]
> really dumb question @mjlbach :
> 
> is pyright supposed to work as an alternative to pyls now?

Fom my understanding this should currently be the best python language server out there.

**Comment by RonanMacF** [comment]
Anyone manage to get this setup using the code pasted by OP above? I have the pyright installed and can see that I have the pyright object selected but when I try run setup i'm running into `attempt to call field 'setup' (a nil value)`

**Comment by mjlbach** [comment]
This was merged, did you try just using the current plugin? And I use the snippet I posted

**Comment by RonanMacF** [comment]
I am actually using your nix neovim-nightly-master which hasn't updated yet but was eager to try it out :) Yep copy + paste just changed `local servers = { 'pyright' }` and removed the diagnostic line. I'll wait for build to update and try it again though

**Comment by mjlbach** [comment]
Ah that might be out of date date, you may need to try replace nvim_lap with lspconfig in the snippet

**Comment by YodaEmbedding** [comment]
> > @mjlbach excited to try this. does it still need [neovim/neovim#12638](https://github.com/neovim/neovim/pull/12638) in order to work? if so, crossed_fingers that gets merged soon!
> 
> @kamalmarhubi Nope, I added a [temporary fix](https://github.com/neovim/nvim-lspconfig/pull/303/files#diff-746dacea82fe0b83d0393aa4539dee7b607672f64d6fd4454f19e6de6d881a86R21-R27) to bypass the need for that PR, I'll remove the fix once that PR is merged.

Since the PR was merged, is the temporary fix still needed?

**Comment by mjlbach** [comment]
Pyright (as seen in the quickstart guide), works fine. No workarounds needed.

**Comment by YodaEmbedding** [comment]
Sorry, I was looking at the wrong commit.

**Pull request merged**
